### DISCUSSION TITLE
Cholesky decomposition performed for every sample(?)

### DISCUSSION THREAD

**Discussion by jonny-so** [blackjax-devs/blackjax#758] [General]
Hi, unless I have missed something, it looks as though (e.g.) the NUTS sampler implementation creates a new Euclidean metric object for every sample https://github.com/blackjax-devs/blackjax/blob/65ae00e95aa122a581caef2d6eb1949cafbf8905/blackjax/mcmc/nuts.py#L123 which in turn performs a Cholesky decomposition of the (inverse) mass matrix. Is that correct? It should only be necessary to do this at the end of the adaptation windows. In principle the compiler could be eliding the redundant calls, but I would guess that is asking too much of it in this case.

**Comment by junpenglao** [comment]
Not sure i understand, that line is a non-op for NUTS because `inverse_mass_matrix` would be a `Metric` type

**Reply by jonny-so** [reply]
It is not, as far as I can tell. The window adaptation returns an array, and so unless the user converts it to a metric before calling `blackjax.nuts`, it remains an array. It is possible I'm missing something still, but the cholesky code is getting hit for every sample when I run the example from the quickstart page (modified to pass `is_mass_matrix_diagonal=False` to `window_adaptation`). The same thing happens during warmup too. I guess `window_adaptation` should store a Metric rather than an Array, but the fix is not quite that simple, because Metric is not a jax type, and so can't be used as e.g. part of a scan carry state.

**Reply by junpenglao** [reply]
ah i see you meant during warmup - it wouldnt be every sample because the new mass matrix only switch at the end of the slow window. IIUC this is unavoidable anyway 
cc @AdrienCorenflos to confirm.

**Reply by jonny-so** [reply]
No it happens both during warmup and after, for every sample. I just meant a potential fix is to have warmup_adaptation store a metric as opposed to an array, and update it at the end of every window. The issue is that the nuts kernel is getting called with an array inverse_mass_matrix for every sample, which means it keeps creating new metric objects.

**Reply by junpenglao** [reply]
OK i think i see what you mean, I think adding another 
```
inverse_mass_matrix = metrics.default_metric(inverse_mass_matrix)
```
in 

https://github.com/blackjax-devs/blackjax/blob/65ae00e95aa122a581caef2d6eb1949cafbf8905/blackjax/mcmc/nuts.py#L150

should fix it?

**Reply by jonny-so** [reply]
I think it would fix it for the top level API, but not during warmup which calls `kernel()` directly. Maybe it makes sense to add a check inside `kernel` that it is in fact a metric object, to avoid a silent performance pitfall.  It looks like barker.py also has the same issue.

Fixing it during warmup looks a little trickier. Ideally `WindowAdaptationState` would store `inverse_mass_matrix` as a Metric rather than an Array, but Metric is not a jax type because it contains various function-valued fields.

**Reply by jonny-so** [reply]
`hmc` and `dynamic_hmc` also actually. One option to resolve this for window adaptation would be to have separate `EuclideanMetric` and `RiemannianMetric` types, with the former being a proper jax type (so it can be used inside scan), as only the latter really needs to contain closures. Then the `Metric` type would be removed altogether, and `MetricTypes` would be

`MetricTypes = Union[EuclideanMetric, Array, RiemannianMetric, Callable[[ArrayLikeTree], Array]]`

or going further, maybe it makes sense to disallow `Array` and `Callable` being used as metrics directly. e.g. it may be cleaner to have

`MetricTypes = Union[EuclideanMetric, RiemannianMetric]`

and only allow explicit construction of those types from `Array`/`Callable`, rather than allowing a kind of implicit conversion. This is just one suggestion -- if you have a preferred way to fix the window adaptation I'm happy to do it and submit a PR.

**Reply by junpenglao** [reply]
Thanks for the thoughtful reply! I agree `MetricTypes = Union[EuclideanMetric, RiemannianMetric]` is a nice solution, but is there a way to extend to other API that depends on `Metric` as well?

**Reply by junpenglao** [reply]
I see that there is 2 route:
1. low level api should only accept a `MetricTypes`
2. play with hashing so if the input array is identical we can reuse the converted `Metric`.

**Comment by junpenglao** [comment]
Converted to an issue https://github.com/blackjax-devs/blackjax/issues/760

**Comment by AdrienCorenflos** [comment]
Sorry this took me a bit of time to look at (vacation etc). I don't really like the idea of implementing a solution to a problem we have not confirmed exists.

I'm not 100% confident what regime we are worried about though so I can't really go in and check: can @jonny-so provide an example (JITTED) where this is happening and shouldn't. I don't really mind if this is a minimal example or a convoluted for this one, just maybe link to a colab/repo where this is happening and it can be confirmed by using JAX debug calls in the stack. If you can then explain with the example why you think the recomputation of the cholesky shouldn't be happening it would be even better.

**Reply by AdrienCorenflos** [reply]
To be clear, I'm happy to go in and debug once you give me a starting point, there is no need to isolate the issue yourself.

**Reply by jonny-so** [reply]
Sure, I've forked the repo at https://github.com/jonny-so/blackjax-issue760. I added a single debug trace to metrics.py, and added a small example of the problem in test.py -- changes can be seen here https://github.com/jonny-so/blackjax-issue760/commit/c7a330e7a89e1d0294d4a9b3af831dfa1d63e175.

If you run (e.g.):
```bash
git clone https://github.com/jonny-so/blackjax-issue760
cd blackjax-issue760
python -m venv venv
source venv/bin/activate
pip install -e .
python test.py
```

then you should see something like:
```
warming up
cholesky!
... (repeated 200 times)
warmed up, pos={'loc': Array(8.949477, dtype=float32), 'log_scale': Array(3.0040097, dtype=float32)}
cholesky!
... (repeated 100 times)
```

You can ignore the "pos={..}" part of the "warmed up" line, I just included that to force the print traces to happen in the right order.

If you look in `test.py`, you can see that I have copied the example given in the quickstart page of the blackjax docs, with 2 modifications, 1) I have used `is_mass_matrix_diagonal=False`, and 2) I have wrapped the whole thing in `jax.jit`.  The code uses a window adpatation with 200 steps, and then draws 100 samples using the fitted parameters.  The mass matrix Cholesky should only need to happen a few times during the window adaptation phase (at the end of "middle" adaptation windows, when the mass matrix actually changes), and shouldn't happen at all after warmup. The example shows it is happening for every sample, both during and after warmup.

Let me know if there is anything else I can do.
